go/vt/vtgate/planbuilder/join.go

```diff
@@ -98,22 +111,126 @@ func newJoin(lpb, rpb *primitiveBuilder, ajoin *sqlparser.JoinTableExpr) error {
 			return errors.New("unsupported: join with USING(column_list) clause")
 		}
 	}
-	lpb.bldr = &join{
-		weightStrings: make(map[*resultColumn]int),
-		Left:          lpb.bldr,
-		Right:         rpb.bldr,
-		ejoin: &engine.Join{
-			Opcode: opcode,
-			Vars:   make(map[string]int),
-		},
+
+	joinStruct, err := createJoinStruct(lpb, rpb, opcode, useHashJoin, ajoin)
+	if err != nil {
+		return err
 	}
+	lpb.bldr = joinStruct
 	lpb.bldr.Reorder(0)
-	if ajoin == nil || opcode == engine.LeftJoin {
+
+	if ajoin == nil || opcode == engine.LeftJoin || useHashJoin {
 		return nil
 	}
 	return lpb.pushFilter(ajoin.Condition.On, sqlparser.WhereStr)
 }
 
+func createJoinStruct(lpb, rpb *primitiveBuilder, opcode engine.JoinOpcode, useHashJoin bool, ajoin *sqlparser.JoinTableExpr) (*join, error) {
+	j := join{
+		weightStrings: make(map[*resultColumn]int),
+		Left:          lpb.bldr,
+		Right:         rpb.bldr,
+		Vars:          make(map[string]int),
+		Opcode:        opcode,
+	}
+
+	if useHashJoin {
+		if ajoin == nil {
+			return nil, vterrors.New(vtrpc.Code_INTERNAL, "need an ajoin to create a hash join")
+		}
+		comparisons, err := extractJoinComparisons(ajoin.Condition.On)
+		if err != nil {
+			return nil, err
+		}
+
+		j.joinImpl = &hashJoin{comparisonExprs: comparisons}
+	} else {
+		j.joinImpl = &nestedLoopJoin{}
+	}
+	return &j, nil
+}
+
+var _ joinStyle = (*hashJoin)(nil)
+
+// extractJoinComparisons recursively walks down an AND tree to produce an array with all the equals comparisons
+func extractJoinComparisons(expr sqlparser.Expr) ([]*sqlparser.ComparisonExpr, error) {
+	cmp, isComparison := expr.(*sqlparser.ComparisonExpr)
+	if isComparison {
+		if cmp.Operator != "=" {
+			return nil, vterrors.Errorf(vtrpc.Code_INTERNAL, "can only handle simple equality between columns using hash joins - got %s", cmp.Operator)
+		}
+
+		return []*sqlparser.ComparisonExpr{cmp}, nil
+	}
+
+	and, isAnd := expr.(*sqlparser.AndExpr)
+	if isAnd {
+		lft, err := extractJoinComparisons(and.Left)
+		if err != nil {
+			return nil, err
+		}
+		rgt, err := extractJoinComparisons(and.Right)
+		if err != nil {
+			return nil, err
+		}
+		return append(lft, rgt...), nil
+	}
+	return nil, vterrors.Errorf(vtrpc.Code_INTERNAL, "can only handle one or more equality comparisons between columns using hash joins - got %s", expr)
```

Comment:
I think you should add the fact that all the comparisons must be conjunctive(a bunch of and experessions).